### PR TITLE
Enable zone detection and automatic OSD placement calculation

### PR DESCRIPTION
**Pull request by jarrpa**
This PR allows ocs-operator to dynamically determine a best-effort placement strategy and failureDomain for Rook-Ceph OSDs.

**NOTE:** This changes the behavior of the operator's generation of Rook-Ceph StorageClassDeviceSets. Previously, a number of StorageClassDeviceSets were created equal to a given StorageDeviceSet's Count. Now, the number of StorageClassDeviceSets created is equal to the value of a given StorageDeviceSet's Replica. StorageDeviceSet.Count is now mapped to StorageClassDeviceSet.Count.

Resolves: #107

### CODE REVIEW

**Comment by jarrpa** [comment]
Currently passes CI tests locally. Still need to write proper unit tests. Currently testing on my own AWS environments.

File: pkg/controller/storagecluster/reconcile.go [hunk 1]
```diff
@@ -23,15 +26,19 @@ import (
...
-	corev1 "k8s.io/api/core/v1"
 )
 
 const (
 	nodeAffinityKey   = "cluster.ocs.openshift.io/openshift-storage"
 	nodeTolerationKey = "node.ocs.openshift.io/storage"
 )
 
-var defaultMonCount int = 3
+var validTopologyLabelKeys = []string{
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go]: Can we make use of keys from https://github.com/kubernetes/api/blob/master/core/v1/well_known_labels.go instead?

File: pkg/controller/storagecluster/reconcile.go [hunk 2]
```diff
@@ -275,6 +290,55 @@ func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcil
 	return reconcile.Result{}, phaseErr
 }
 
+// getNodeTopologyMap
+func (r *ReconcileStorageCluster) getNodeTopologyMap(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go]: If this function is expected to write topologyMap spec to server then this should probably be renamed to setNodeTopologyMap?

Why do we have to pass reqlogger? ReconcileStorageCluster already has that as a property if needed.

File: pkg/controller/storagecluster/reconcile.go [hunk 3]
```diff
@@ -23,15 +26,19 @@ import (
...
 	nodeTolerationKey = "node.ocs.openshift.io/storage"
 )
 
-var defaultMonCount int = 3
+var validTopologyLabelKeys = []string{
+	"failure-domain.beta.kubernetes.io",
+	"failure-domain.kubernetes.io",
+}
+
+var defaultMonCount = 3
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go]: Why not just assign it to monCount? Unless we are allowing it to be configurable, which is not the case currently.

File: pkg/controller/storagecluster/reconcile.go [hunk 4]
```diff
@@ -275,6 +290,55 @@ func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcil
...
+					if !topologyMap.Contains(label, value) {
+						topologyMap.Add(label, value)
+					}
+				}
+			}
+		}
+
+	}
+
+	sc.Status.NodeTopologies = topologyMap
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go]: This will just update the object and not write it to the server. (if that is intended)
Use `Client().Status().Update()` instead to update object definition on server.

File: pkg/controller/util/status.go [hunk 5]
```diff
@@ -12,18 +12,18 @@ import (
 // These constants represent the overall Phase as used by .Status.Phase
 var (
 	// PhaseIgnored is used when a resource is ignored
-	PhaseIgnored string = "Ignored"
+	PhaseIgnored = "Ignored"
```
Comment (umangachapagain) [inline pkg/controller/util/status.go]: These changes should at least be a separate commit.

File: pkg/controller/storagecluster/reconcile.go [hunk 6]
```diff
@@ -461,3 +510,57 @@ func newCephCluster(sc *ocsv1.StorageCluster, cephImage string) *cephv1.CephClus
 
 	return cephCluster
 }
+
+// newStorageClassDeviceSets converts a list of StorageDeviceSets into a list of Rook StorageClassDeviceSets
+func newStorageClassDeviceSets(storageDeviceSets []ocsv1.StorageDeviceSet, topologyMap ocsv1.NodeTopologyMap) []rook.StorageClassDeviceSet {
+	var storageClassDeviceSets []rook.StorageClassDeviceSet
+
+	for _, ds := range storageDeviceSets {
+		topologyKey := ds.TopologyKey
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go]: Does `ds.TopologyKey` exist? TopologyKey of which spec will this pick?

File: pkg/apis/ocs/v1/topologymap.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	if _, ok := m.Labels[topologyKey]; !ok {
+		m.Labels[topologyKey] = TopologyLabelValues{}
+	}
+
+	m.Labels[topologyKey][value] = true
+}
+
+// GetKeyValues returns a node label matching the topologyKey and all values
+// for that label across all storage nodes
+func (m *NodeTopologyMap) GetKeyValues(topologyKey string) (string, []string) {
```
Comment (umangachapagain) [inline pkg/apis/ocs/v1/topologymap.go]: `GetNodeLabelUsingKey` looks understandable and self-explanatory.

File: pkg/controller/storagecluster/reconcile.go (hunk 3 above)
Comment (davidvossel) [reply pkg/controller/storagecluster/reconcile.go]: I had made the mon count configurable using an env var a couple of weeks ago.  I needed a way to override the default of '3' in order to support replica 2 for the test suite. 

If you all want to remove usage of the MON_COUNT_OVERRIDE variable now, that should be fine because the e2e tests now run with replica 3. The variable does serve as an escape hatch for us though if we never need to configure the mon count for some reason.

File: pkg/controller/storagecluster/reconcile.go (hunk 4 above)
Comment (davidvossel) [reply pkg/controller/storagecluster/reconcile.go]: the storagecluster is updated once at the end of the reconcile loop

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 8]
```diff
@@ -58,6 +75,23 @@ type StorageClusterStatus struct {
...
+// TopologyLabelValues is a map of strings
+//
+// The current implementation facilitates searching for a specific value
+// without having to loop through all values
+type TopologyLabelValues map[string]bool
+
+// NodeTopologyMap represents the list of all values of all topology labels
+// across all nodes in the StorageCluster
+type NodeTopologyMap struct {
+	Labels map[string]TopologyLabelValues `json:"labels"`
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/apis/ocs/v1/storagecluster_types.go]: Yup.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 9]
```diff
@@ -58,6 +75,23 @@ type StorageClusterStatus struct {
...
+	// the StorageCluster's placement selector.
+	// +optional
+	NodeTopologies NodeTopologyMap `json:"nodeTopologies,omitempty"`
+}
+
+// TopologyLabelValues is a map of strings
+//
+// The current implementation facilitates searching for a specific value
+// without having to loop through all values
+type TopologyLabelValues map[string]bool
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/apis/ocs/v1/storagecluster_types.go]: Indeed, this is largely a throwaway value. However, having it always be "true" is a minor convenience in the current implementation.

File: pkg/controller/storagecluster/reconcile.go [hunk 10]
```diff
@@ -275,6 +290,55 @@ func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcil
...
+		return err
+	}
+
+	topologyMap := ocsv1.NodeTopologyMap{}
+
+	for _, node := range nodes.Items {
+		labels := node.Labels
+		for _, label := range labels {
+			for _, key := range validTopologyLabelKeys {
+				if strings.Contains(label, key) {
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Because `validTopologyLabelKeys` is something like `failure-domain.kubernetes.io` and not the full label, e.g. `failure-domain.kubernetes.io/zone`.

File: pkg/controller/storagecluster/reconcile.go [hunk 11]
```diff
@@ -141,6 +148,13 @@ func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcil
 		}
 	}
 
+	// Get storage node topology labels
+	err = r.getNodeTopologyMap(instance, reqLogger)
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Any suggestions?

File: pkg/apis/ocs/v1/topologymap.go (hunk 7 above)
Comment (jarrpa) [reply pkg/apis/ocs/v1/topologymap.go]: You're not just getting a label, you are getting all values for all labels matching the specified `topologyKey`.

File: pkg/controller/storagecluster/reconcile.go (hunk 1 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Not strictly, since we'll want to provide our own labels in the future. Also I'm primarily interested in the left-hand side of the label and not the right.

File: pkg/controller/storagecluster/reconcile.go (hunk 2 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: You can't set what you don't have. This is very specifically getting information and then saving it somewhere.

As for reqLogger, I was just blindly following conventions elsewhere in the file. :) I can change that.

File: pkg/controller/storagecluster/reconcile.go (hunk 6 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: It does. Defaults to empty string.

File: pkg/controller/storagecluster/reconcile.go [hunk 12]
```diff
@@ -460,3 +510,57 @@ func newCephCluster(sc *ocsv1.StorageCluster, cephImage string) *cephv1.CephClus
...
+						topologyIndex = topologyIndex - len(topologyKeyValues)
+					}
+					nodeZoneSelector := corev1.NodeSelectorRequirement{
+						Key:      topologyKey,
+						Operator: corev1.NodeSelectorOpIn,
+						Values:   []string{topologyKeyValues[topologyIndex]},
+					}
+					placement.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions = append(placement.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions, nodeZoneSelector)
+				} else {
+					portable = false
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Right, this does that. But if you have three or more of your given topology key (e.g. `zone`) then you can set portable to true.

File: pkg/controller/util/status.go (hunk 5 above)
Comment (jarrpa) [reply pkg/controller/util/status.go]: Everything should be in proper commits. :) That's coming later.

File: pkg/controller/storagecluster/reconcile.go (hunk 1 above)
Comment (rohantmp) [reply pkg/controller/storagecluster/reconcile.go]: I think what umanga means is to replace the two literals here with references to the consts in that file.

File: pkg/controller/storagecluster/reconcile.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (rohantmp) [reply pkg/controller/storagecluster/reconcile.go]: r.updateNodeTopologyStatus?

File: pkg/controller/storagecluster/reconcile.go (hunk 1 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Yes, and I pointedly don't want to for the reasons listed above. :)

File: pkg/controller/storagecluster/reconcile.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Works for me

File: pkg/controller/storagecluster/reconcile.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (rohantmp) [reply pkg/controller/storagecluster/reconcile.go]: I think we should put these consts in validTopologyLabelKeys: https://github.com/kubernetes/api/blob/master/core/v1/well_known_labels.go#L21

If  we want to support arbitrary topology keys, we should have restrictive regex for them. Allowing this might have unintended consequences from label collisions. Infra labels should either come as documented behavior from openshift/kubernetes, or be prefixed with the ocs domain.

File: pkg/controller/storagecluster/reconcile.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (rohantmp) [reply pkg/controller/storagecluster/reconcile.go]: oh, I hadn't looked at the validTopologyKeys and assumed the hostname key was in there.

**Comment by jarrpa** [comment]
New updates pushed. I managed to get basic happy path testing to succeed with 3 AZs and split the wip commit into more logical and reasonable commits. Things outstanding:

* Test for fewer than 3 AZs
* Write unit tests

File: pkg/controller/storagecluster/reconcile.go [hunk 13]
```diff
@@ -409,3 +456,63 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+
+		topologyKey := ds.TopologyKey
+		topologyKeyValues := []string{}
+		noPlacement := ds.Placement.NodeAffinity == nil && ds.Placement.PodAffinity == nil && ds.Placement.PodAntiAffinity == nil
+
+		if topologyKey != "" && noPlacement {
+			topologyKey, topologyKeyValues = topologyMap.GetKeyValues(topologyKey)
+		}
+
+		for i := 0; i <= ds.Replica; i++ {
```
Comment (davidvossel) [inline pkg/controller/storagecluster/reconcile.go]: just double checking this is right. or should it be ```i < ds.Replica``` ?

File: pkg/controller/storagecluster/reconcile.go [hunk 14]
```diff
@@ -409,3 +456,63 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+			portable := ds.Portable
+
+			if noPlacement {
+				placement = defaults.DaemonPlacements["osd"]
+				if len(topologyKeyValues) >= 3 {
+					portable = true
+					placement.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution[0].PodAffinityTerm.TopologyKey = topologyKey
+
+					topologyIndex := i
+					for topologyIndex >= len(topologyKeyValues) {
```
Comment (davidvossel) [inline pkg/controller/storagecluster/reconcile.go]: Would it make more sense to use the modulus here to wrap around the index when there are more ds than topologykeyvalues?

File: pkg/controller/storagecluster/reconcile.go [hunk 15]
```diff
@@ -409,3 +456,63 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+		}
+
+		for i := 0; i <= ds.Replica; i++ {
+			placement := ds.Placement
+			portable := ds.Portable
+
+			if noPlacement {
+				placement = defaults.DaemonPlacements["osd"]
+				if len(topologyKeyValues) >= 3 {
+					portable = true
```
Comment (davidvossel) [inline pkg/controller/storagecluster/reconcile.go]: out of curiosity here. why isn't portable a reflection of whether the storageclass being used is portable or not rather than how many AZs exist?

File: hack/gen-deployment-yaml.sh [hunk 16]
```diff
@@ -5,7 +5,7 @@ set -e
 source hack/common.sh
 
 # use the default image name when generating deploy-with-olm.yaml
-DEPLOY_YAML_IMAGE_NAME="${DEFAULT_IMAGE_REGISTRY}/${DEFAULT_REGISTRY_NAMESPACE}/${IMAGE_NAME}:${DEFAULT_IMAGE_TAG}"
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply hack/gen-deployment-yaml.sh]: That is exactly what I want to do. I want this to generate something that points to an image under my account/namespace for personal testing. See here: https://github.com/jarrpa/ocs-operator/commit/2281e933b571e3f7f32101692a4512f4f388cc88

File: pkg/controller/storagecluster/reconcile.go (hunk 13 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: ....it is not right! Good catch.

File: pkg/controller/storagecluster/reconcile.go (hunk 14 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Not a bad idea, I completely forgot about it!

File: pkg/controller/storagecluster/reconcile.go (hunk 15 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Because, at the moment, we don't want to support portable OSDs without having at least three AZs, since we could end up clustering all OSDs of a given PG onto the same node via node failures.

File: pkg/controller/storagecluster/reconcile.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (rohantmp) [reply pkg/controller/storagecluster/reconcile.go]: Maybe it would be clearer to say I don't want to hijack the `failure-domain.kubernetes.io/xxx` with ocs-specific behaviour. Someone else might have the same idea and cause a conflict. If we are planning to allow that, we should do `failure-domain..ocs.openshift.io/xxx`

File: hack/gen-deployment-yaml.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (davidvossel) [reply hack/gen-deployment-yaml.sh]: ah, okay. makes sense. Just don't accidentally commit the deploy-with-olm.yaml pointing to your build I guess.

We probably need a ci check to guarantee the ocs-registry image in the deploy-with-olm.yaml that's committed to the src tree is coming from the quay.io/ocs-dev/ registry so this doesn't accidentally get pointed to a developers personal registry.

File: pkg/controller/storagecluster/reconcile.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Yeah, that's fine. We're just working with what we got for now.

**Comment by jarrpa** [comment]
This PR has been updated to include some basic unit testing. Manual testing has also verified this behaves as expected in AWS with two and three AZs. I added the following note to the PR description as well:

**NOTE:** This changes the behavior of the operator's generation of Rook-Ceph StorageClassDeviceSets. Previously, a number of StorageClassDeviceSets were created equal to a given StorageDeviceSet's Count. Now, the number of StorageClassDeviceSets created is equal to the value of a given StorageDeviceSet's Replica. StorageDeviceSet.Count is now mapped to StorageClassDeviceSet.Count.

**Comment by jarrpa** [comment]
@kshlm @davidvossel PTAL

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 17]
```diff
@@ -25,16 +25,34 @@ type StorageClusterSpec struct {
...
-	Count           int                          `json:"count"`
-	Resources       corev1.ResourceRequirements  `json:"resources"`
-	Placement       rookalpha.Placement          `json:"placement"`
+	Count int `json:"count"`
+
+	// Replica is the number of StorageClassDeviceSets for this
+	// StorageDeviceSet
+	// +kubebuilder:validation:Minimum=1
+	// +optional
+	Replica int `json:"replica,omitempty"`
```
Comment (davidvossel) [inline pkg/apis/ocs/v1/storagecluster_types.go]: Can we start using this new `Replica` option in the default test StorageCluster that our functional tests deploy and test with?

That can be found here, ```pkg/deploy-manager/storagecluster.go``` in the ```defaultStorageCluster()``` function. Making this change will ensure we can deploy using a config that more closely resembles what we're going with for production. 

You might also consider updating the ```deploy/crds/ocs_v1_storagecluster_cr.yaml``` example CR.

File: pkg/controller/storagecluster/reconcile.go [hunk 18]
```diff
@@ -409,3 +466,80 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+		// This is a temporary hack in place due to limitations in
+		// the current implementation of the OCP console.
+		// TODO: Remove this behavior when OCP console is updated
+		count := ds.Count / 3
+
+		for i := 0; i < replica; i++ {
+			placement := rook.Placement{}
+			portable := ds.Portable
+
+			if noPlacement {
```
Comment (davidvossel) [inline pkg/controller/storagecluster/reconcile.go]: I think it would be worthwhile to document some of this default Placement behavior in the top level ocs-operator README.md file soon.

File: pkg/apis/ocs/v1/storagecluster_types.go (hunk 17 above)
Comment (jarrpa) [reply pkg/apis/ocs/v1/storagecluster_types.go]: I'd like to keep the default test StorageCluster reflecting the current state of the product. Since the OCP console is currently stuck using a Count of 3 and a Replica of 0 (default 3), we should keep it that way until the console is updated.

File: pkg/controller/storagecluster/reconcile.go [hunk 19]
```diff
@@ -409,3 +466,80 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+
+		replica := ds.Replica
+		if replica == 0 {
+			replica = defaults.DeviceSetReplica
+		}
+
+		// This is a temporary hack in place due to limitations in
+		// the current implementation of the OCP console.
+		// TODO: Remove this behavior when OCP console is updated
+		count := ds.Count / 3
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Done.

File: pkg/controller/storagecluster/reconcile.go (hunk 18 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: I'll do it closer to release, once we're sure the placement logic has stabilized.

File: pkg/controller/storageclusterinitialization/storageclusterinitialization_controller.go [hunk 20]
```diff
@@ -472,7 +472,7 @@ func (r *ReconcileStorageClusterInitialization) newCephBlockPoolInstances(initDa
 				Namespace: initData.Namespace,
 			},
 			Spec: cephv1.PoolSpec{
-				FailureDomain: "host",
+				FailureDomain: initData.Spec.FailureDomain,
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storageclusterinitialization/storageclusterinitialization_controller.go]: I'm not sure. Since we're effectively just passing a variable from the InitData to this field, does it make sense to just have a test for setting values without a calculation?

File: pkg/controller/storagecluster/reconcile.go (hunk 18 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: Created Issue #192.

Comment (davidvossel) [reply pkg/controller/storagecluster/reconcile.go]: excellent, thanks

File: pkg/apis/ocs/v1/storagecluster_types.go (hunk 17 above)
Comment (davidvossel) [reply pkg/apis/ocs/v1/storagecluster_types.go]: Somehow we need to functionally prove this configuration works. Otherwise we run the risk of breaking it. 

do we need to restructure the functional tests to allow us to cycle through multiple StorageCluster configurations?

Comment (jarrpa) [reply pkg/apis/ocs/v1/storagecluster_types.go]: Simply testing that the expected number of OSDs gets created should suffice for now. I don't think we should spend time testing multiple StorageCluster configurations since we will only support one for the initial release.

File: pkg/controller/storageclusterinitialization/storageclusterinitialization_controller.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storageclusterinitialization/storageclusterinitialization_controller.go]: This was already technically being taken care of by the unit tests. I added an explicit configuration of `failureDomain` to the scinit objects just to be sure.

**Comment by jarrpa** [comment]
Looks like the functional tests are failing on creating the OCP cluster... guess we'll have to wait for that to clear up, I'd definitely not want to merge without them running.

**Comment by davidvossel** [comment]
> Looks like the functional tests are failing on creating the OCP cluster... guess we'll have to wait for that to clear up, I'd definitely not want to merge without them running.

yeah, the issue is out of our control right now. It should be impacting everyone using openshift/release

**Comment by jarrpa** [comment]
@davidvossel CI passed!

**Comment by davidvossel** [comment]
yay, our functests pass.

I re-kicked off the ocs-ci e2e tests, but i'm not necessarily expecting that test suite to pass here. There's some flakiness there we're sorting through. I just wanted to see it run since the last run was cut short by an openshift/release related issue.

i'm doing a final run through this pr now.

File: pkg/controller/storagecluster/reconcile.go [hunk 21]
```diff
@@ -409,3 +466,85 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+
+		// This is a temporary hack in place due to limitations in
+		// the current implementation of the OCP console. The console
+		// is hardcoded to create a StorageCluster with a Count of 3,
+		// as made sense for the previous behavior, but it cannot be
+		// updated until the next z-stream release of OCP 4.2. This
+		// workaround is to enable the new behavior while the console
+		// is waiting to be updated.
+		// TODO: Remove this behavior when OCP console is updated
+		count := ds.Count / 3
```
Comment (davidvossel) [inline pkg/controller/storagecluster/reconcile.go]: This workaround is my only remaining reservation about this PR. I'm wondering if there's a safer way for us to express this workaround that won't inadvertently conflict with future configs. 

For example, as it is here if we set ds.Replica = 3 and ds.Count = 3, this work around would only give us 3 osds instead of 6.

What if we just checked to see if ds.Replica == 0 as the condition for this workaround. If Replica = 0, then attempt to put each "Count" on an individtual StorageClassDeviceSet

It wouldn't be a workaround anymore at that point. If you want the behavior of 1 replica with multiple OSDs, then you'll need to use Replica and Count.  If you only specify Count, then you're getting a StorageClassDeviceSet for each Count by default. 

Does that make sense?

Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: It does. I've moved the workaround into the `if replica == 0` stanza.

File: pkg/controller/storagecluster/reconcile.go [hunk 22]
```diff
@@ -409,3 +466,87 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+			// This is a temporary hack in place due to limitations
+			// in the current implementation of the OCP console.
+			// The console is hardcoded to create a StorageCluster
+			// with a Count of 3, as made sense for the previous
+			// behavior, but it cannot be updated until the next
+			// z-stream release of OCP 4.2. This workaround is to
+			// enable the new behavior while the console is waiting
+			// to be updated.
+			// TODO: Remove this behavior when OCP console is updated
+			count = count / 3
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: I'm still somewhat concerned with the change of meaning of `count` in the API (that this workaround is undoing for the case of `replica==0`). Unless we are managing to get the change (corresponding to the removal of the hack) into the UI before our first OCS release, there's always going to be a time window where things are broken (since we can not release OCP and OCS in perfect sync).

Of course, having this hack under the condition of `replica == 0` is pretty clever in that it allows us to change the UI to use replica explicitly, and the operator would have the new logic there right away. And later we could get rid of the hack...

Just for completeness, the alternative that I discussed with @jarrpa earlier today would be leave the meaning of `count` to be the total number of disks and distribute this number across the zones (or, later, racks). It would avoid breaking API changes. But I am not 100% sure what I would consider to be the more natural API here...

File: pkg/controller/storagecluster/reconcile.go [hunk 23]
```diff
@@ -301,6 +361,22 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1.StorageCluster, re
...
 }
 
+// determineFailureDomain determines the appropriate Ceph failure domain based
+// on the storage cluster's topology map
+func determineFailureDomain(topologyMap *ocsv1.NodeTopologyMap) string {
+	failureDomain := "host"
+
+	for label, labelValues := range topologyMap.Labels {
+		if strings.Contains(label, "zone") {
+			if len(labelValues) >= 3 {
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: At some point it would be great to replace the hard-coded `3` here with the replica number from the storageDeviceSet.

File: pkg/controller/storagecluster/reconcile.go (hunk 22 above)
Comment (davidvossel) [reply pkg/controller/storagecluster/reconcile.go]: > Just for completeness, the alternative that I discussed with @jarrpa earlier today would be leave the meaning of count to be the total number of disks and distribute this number across the zones (or, later, racks). It would avoid breaking API changes. But I am not 100% sure what I would consider to be the more natural API here...

@obnoxxx Is the motivation around keeping count to mean total disks driven by maintaining API compatibility or because there's a use case where someone would need to explicitly control disk count and zones/racks count independently?
